### PR TITLE
Spike/basic support for alarms and conditions

### PR DESCRIPTION
**Pull request by htbmw**

### CODE REVIEW

File: src/Application/Factories/AlarmMonitoredItemServiceInitializerFactory.cs [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+using OMP.Connector.Domain.Schema;
+using Opc.Ua.Client;
+
+namespace OMP.Connector.Application.Factories
+{
+    public class AlarmMonitoredItemServiceInitializerFactory : IAlarmMonitoredItemServiceInitializerFactory
+    {
+        private readonly IServiceProvider _serviceProvider;
+
+        public AlarmMonitoredItemServiceInitializerFactory(IServiceProvider serviceProvider)
```
Comment (ivomp) [inline src/Application/Factories/AlarmMonitoredItemServiceInitializerFactory.cs:18]: Why do we inject the DI container here? By doing this we do not clearly state to the outside world what is required for this class to function, so the outside world needs to know what is going on inside to be able to use this class. Also problematic for unit test.
Is there an edge case here regarding the timing of external initialization why we do this?

File: src/Application/Factories/AlarmMonitoredItemServiceInitializerFactory.cs [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+        {
+            this._serviceProvider = serviceProvider;
+        }
+
+        public MonitoredItem Initialize(AlarmSubscriptionMonitoredItem monitoredItem, IComplexTypeSystem complexTypeSystem, TelemetryMessageMetadata telemetryMessageMetadata, Session session)
+        {
+            var opcMonitoredItemService = this._serviceProvider.GetRequiredService<IOpcAlarmMonitoredItemService>();
```
Comment (ivomp) [inline src/Application/Factories/AlarmMonitoredItemServiceInitializerFactory.cs:25]: Can we not inject this in the ctor instead of the entire DI container?

File: src/Application/Factories/AlarmSubscriptionProviderFactory.cs [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        private readonly IOptions<ConnectorConfiguration> _connectorConfiguration;
+        private readonly AlarmMonitoredItemServiceInitializerFactoryDelegate _monitoredItemServiceInitializerFactory;
+        private readonly AlarmMonitoredItemValidator _alarmMonitoredItemValidator;
+
+        public AlarmSubscriptionProviderFactory(
+            IAlarmSubscriptionRepository dataManagementService,
+            ILoggerFactory loggerFactory,
+            IOptions<ConnectorConfiguration> connectorConfiguration,
+            AlarmMonitoredItemServiceInitializerFactoryDelegate monitoredItemServiceInitializerFactory,
+            AlarmMonitoredItemValidator alarmMonitoredItemValidator)
```
Comment (ivomp) [inline src/Application/Factories/AlarmSubscriptionProviderFactory.cs:30]: Can this not be an interface? Do we do it as instances in other areas? (Not sure if already broke this rule elsewhere)

File: src/Application/Factories/IAlarmMonitoredItemServiceInitializerFactory.cs [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+namespace OMP.Connector.Application.Factories
+{
+    public interface IAlarmMonitoredItemServiceInitializerFactory
+    {
+        public MonitoredItem Initialize(
+            AlarmSubscriptionMonitoredItem monitoredItem,
+            IComplexTypeSystem complexTypeSystem,
+            TelemetryMessageMetadata telemetryMessageMetadata,
+            Session session);
```
Comment (ivomp) [inline src/Application/Factories/IAlarmMonitoredItemServiceInitializerFactory.cs:17]: Can this not be our wrapper IOpcUaSession object to ease unit testing and reduce tight coupling?
(The opc ua foundation has still not merged the PR in that introduces the ISession abstraction ;( )

File: src/Application/OpcUa/AlarmConversion.cs [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+{
+    internal static class AlarmConversion
+    {
+        private const string ArrayPostfix = "[]";
+
+        #region Conversion methods
+
+        internal static List<object> ConvertToAlarm(object value, IMapper mapper)
+        {
+            //if (IsPrimitiveStringOrNull(value))
```
Comment (ivomp) [inline src/Application/OpcUa/AlarmConversion.cs:28]: Can we delete all the dead code in here?

File: src/Application/Providers/AlarmsConditions/CreateAlarmSubscriptionProvider.cs [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+    public class CreateAlarmSubscriptionProvider : AlarmSubscriptionProvider<CreateAlarmSubscriptionsRequest, CreateAlarmSubscriptionsResponse>
+    {
+        //TODO: add support for alarm subscription restores
+        private readonly IAlarmSubscriptionRepository _subscriptionRepository;
+        private readonly TelemetryMessageMetadata _messageMetadata;
+        private readonly AlarmMonitoredItemValidator _alarmMonitoredItemValidator;
+        private readonly int _batchSize;
+        private readonly AlarmMonitoredItemServiceInitializerFactoryDelegate _opcMonitoredItemServiceInitializerFactory;
+        private readonly Dictionary<string, List<string>> _groupedItemsNotCreated;
+        private AlarmFilterDefinition m_filter;
```
Comment (ivomp) [inline src/Application/Providers/AlarmsConditions/CreateAlarmSubscriptionProvider.cs:43]: The field 'CreateAlarmSubscriptionProvider.m_filter' is never used

File: src/Application/Providers/AlarmsConditions/CreateAlarmSubscriptionProvider.cs [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+
+        public CreateAlarmSubscriptionProvider(
+            IAlarmSubscriptionRepository subscriptionRepository,
+            ILogger<CreateAlarmSubscriptionProvider> logger,
+            IOptions<ConnectorConfiguration> connectorConfiguration,
+            AlarmMonitoredItemServiceInitializerFactoryDelegate opcMonitoredItemServiceInitializerFactory,
+            CreateAlarmSubscriptionsRequest command,
+            TelemetryMessageMetadata messageMetadata,
+            AlarmMonitoredItemValidator alarmMonitoredItemValidator) : base(command, connectorConfiguration, logger)
```
Comment (ivomp) [inline src/Application/Providers/AlarmsConditions/CreateAlarmSubscriptionProvider.cs:52]: Interface? (as per previous comment)

File: src/Application/Providers/AlarmsConditions/CreateAlarmSubscriptionProvider.cs [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+                }
+            }
+            catch (Exception ex)
+            {
+                var error = ex.Demystify();
+                errorMessages.Add($"Bad: Could not create / update alarm subscriptions. {error.Message}");
+            }
+
+            //TODO: Implement subscription restore for alarms. Requires different Dto
+            //if (!base.Settings.DisableSubscriptionRestoreService && !errorMessages.Any())
```
Comment (ivomp) [inline src/Application/Providers/AlarmsConditions/CreateAlarmSubscriptionProvider.cs:99]: Place holder code for upcoming feature?

File: src/Application/Providers/AlarmsConditions/RespondToAlarmEventsProvider.cs [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+                    _ => null
+                };
+                var statusCode = CallMethod(eventAction.SourceNodeId, methodNodeId, eventAction.Comment, eventAction.EventId);
+                if(StatusCode.IsBad(statusCode))
+                    results.Add(GetBadCallResult(statusCode, eventAction.Action, eventAction.Comment));
+            }
+
+            this.Logger.Debug($"Created/Updated alarm subscriptions on Endpoint: [{this.EndpointUrl}]");
+
+            return this.GetStatusMessage(results);
```
Comment (ivomp) [inline src/Application/Providers/AlarmsConditions/RespondToAlarmEventsProvider.cs:45]: Can we rather return Task.FromResult(... so we don net get this warning:
This async method lacks 'await' operators and will run synchronously. Consider using the 'await' operator to await non-blocking 
Or is there a reason why we opted fir this return?

File: src/Application/Services/OpcAlarmMonitoredItemService.cs [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+        {
+            this._logger = logger;
+            this._mapper = mapper;
+            this._connectorConfiguration = connectorConfiguration.Value;
+            this._messageSender = messageSender;
+
+            this.Notification += this.OnNotification;
+        }
+
+        public void Initialize(AlarmSubscriptionMonitoredItem monitoredItemCommand, IComplexTypeSystem complexTypeSystem, TelemetryMessageMetadata messageMetadata, Session session)
```
Comment (ivomp) [inline src/Application/Services/OpcAlarmMonitoredItemService.cs:51]: Is it possible to use our abstraction IOpcUaSession object instead of this concrete Session object?

File: src/Application/Factories/AlarmSubscriptionProviderFactory.cs (hunk 3 above)
Comment (ivomp) [reply src/Application/Factories/AlarmSubscriptionProviderFactory.cs:30]: Or the AbstractValidator<..> to minimize the coupling and ease testing?

File: src/Domain.Schema/Converters/SensorMeasurementConverter.cs [hunk 11]
```diff
@@ -8,16 +8,38 @@
...
         protected override bool LoadAfterCreate => false;
 
         public override void WriteJson(JsonWriter writer, object value, JsonSerializer serializer)
-            => serializer.Serialize(writer, value, value.GetType());
+        {
+            if(value is List<SensorMeasurement> list)
+            {
+                foreach (var item in list)
+                {
+                    if(item.Key.Equals("Value"))
```
Comment (ivomp) [inline src/Domain.Schema/Converters/SensorMeasurementConverter.cs:28]: Is there some code missing here? If not can we remove the empty if block?

File: src/Infrastructure/Kafka/KafkaMessageSender.cs [hunk 12]
```diff
@@ -25,12 +27,13 @@
 
 namespace OMP.Connector.Infrastructure.Kafka
 {
-    public class KafkaMessageSender : IMessageSender//, IDisposable
```
Comment (ivomp) [inline src/Infrastructure/Kafka/KafkaMessageSender.cs:28]: Thanks for removing it 🍡

File: src/Infrastructure/MQTT/MqttMessageSender.cs [hunk 13]
```diff
@@ -51,5 +52,10 @@ public bool SendMessageToConfig(AppConfigDto configuration)
         {
             throw new NotSupportedException();
         }
+
+        public Task SendMessageToAlarmsAsync(AlarmMessage alarmMessage)
+        {
+            throw new NotImplementedException();
```
Comment (ivomp) [inline src/Infrastructure/MQTT/MqttMessageSender.cs:58]: Why do we not send alarms here? Should this not also be on the telemetry channel? Or would another channel be required?

**Review by ivomp — CHANGES_REQUESTED** [review]

File: src/Application/Factories/AlarmMonitoredItemServiceInitializerFactory.cs (hunk 1 above)
Comment (htbmw) [reply src/Application/Factories/AlarmMonitoredItemServiceInitializerFactory.cs]: I agree and I cannot recall the reason why it was implemented this way in the Subscription. 
This pattern was copied from the normal Subscription and it existed before the code was migrated to OMP.

I have removed the unnecessary complexity with the dependency injection.

File: src/Application/OpcUa/AlarmConversion.cs (hunk 5 above)
Comment (htbmw) [reply src/Application/OpcUa/AlarmConversion.cs]: Done

File: src/Application/Providers/AlarmsConditions/RespondToAlarmEventsProvider.cs (hunk 9 above)
Comment (htbmw) [reply src/Application/Providers/AlarmsConditions/RespondToAlarmEventsProvider.cs:45]: I removed the async modifier and used the FromResult as suggested.

File: src/Application/Services/OpcAlarmMonitoredItemService.cs (hunk 10 above)
Comment (htbmw) [reply src/Application/Services/OpcAlarmMonitoredItemService.cs]: Done

File: src/Infrastructure/MQTT/MqttMessageSender.cs (hunk 13 above)
Comment (htbmw) [reply src/Infrastructure/MQTT/MqttMessageSender.cs:58]: Must have overlooked it as I initially developed the alarms feature whilst using a kafka application host to test it. :)

File: src/Application/Providers/AlarmsConditions/CreateAlarmSubscriptionProvider.cs (hunk 8 above)
Comment (htbmw) [reply src/Application/Providers/AlarmsConditions/CreateAlarmSubscriptionProvider.cs:92]: Yes correct. Similar to the regular subscription restore, we will provide alarm subscription restore as well.
Will be done in a later story.

File: src/Domain.Schema/Converters/SensorMeasurementConverter.cs (hunk 11 above)
Comment (htbmw) [reply src/Domain.Schema/Converters/SensorMeasurementConverter.cs]: I don't think this is a change that is supposed to go into this PR.
I have removed the change.

File: src/Infrastructure/MQTT/MqttMessageSender.cs (hunk 13 above)
Comment (htbmw) [reply src/Infrastructure/MQTT/MqttMessageSender.cs]: Added

File: src/Application/Factories/AlarmSubscriptionProviderFactory.cs (hunk 3 above)
Comment (htbmw) [reply src/Application/Factories/AlarmSubscriptionProviderFactory.cs]: Although the method was removed as part of another commit based on another code review comment, I did implement the abstraction on field declarations, constructors and some other method signatures.

**Review by ivomp — APPROVED** [review]
